src/native/crt.c

```diff
@@ -44,6 +44,105 @@ void aws_jni_throw_runtime_exception(JNIEnv *env, const char *msg, ...) {
+    jclass cls = (*env)->FindClass(env, "java/nio/ByteBuffer");
+    assert(cls);
+
+    // FindClass() returns local JNI references that become eligible for GC once this native method returns to Java.
+    // Call NewGlobalRef() so that this class reference doesn't get Garbage collected.
+    s_java_byte_buffer.cls = (*env)->NewGlobalRef(env, cls);
+
+    s_java_byte_buffer.get_capacity = (*env)->GetMethodID(env, cls, "capacity", "()I");
+    assert(s_java_byte_buffer.get_capacity);
+
+    s_java_byte_buffer.get_limit = (*env)->GetMethodID(env, cls, "limit", "()I");
+    assert(s_java_byte_buffer.get_limit);
+
+    s_java_byte_buffer.set_limit = (*env)->GetMethodID(env, cls, "limit", "(I)Ljava/nio/Buffer;");
+    assert(s_java_byte_buffer.set_limit);
+
+    s_java_byte_buffer.get_position = (*env)->GetMethodID(env, cls, "position", "()I");
+    assert(s_java_byte_buffer.get_position);
+
+    s_java_byte_buffer.set_position = (*env)->GetMethodID(env, cls, "position", "(I)Ljava/nio/Buffer;");
+    assert(s_java_byte_buffer.set_position);
+
+    s_java_byte_buffer.get_remaining = (*env)->GetMethodID(env, cls, "remaining", "()I");
+    assert(s_java_byte_buffer.get_remaining);
+
+    s_java_byte_buffer.wrap = (*env)->GetStaticMethodID(env, cls, "wrap", "([B)Ljava/nio/ByteBuffer;");
+    assert(s_java_byte_buffer.wrap);
+}
+
+/**
+ * Converts a Java byte[] to a Native aws_byte_cursor
+ */
+struct aws_byte_cursor aws_jni_byte_cursor_from_jbyteArray(JNIEnv *env, jbyteArray array) {
+
+    jboolean isCopy;
+    jbyte *data = (*env)->GetByteArrayElements(env, array, &isCopy);
+    jsize len = (*env)->GetArrayLength(env, array);
+    return aws_byte_cursor_from_array((const uint8_t *)data, (size_t)len);
+}
+
+/**
+ * Converts a Native aws_byte_cursor to a Java byte[]
+ */
+jbyteArray aws_jni_byte_array_from_cursor(JNIEnv *env, const struct aws_byte_cursor *native_data) {
+    jbyteArray jArray = (*env)->NewByteArray(env, native_data->len);
+    (*env)->SetByteArrayRegion(env, jArray, (jsize)0, (jsize)native_data->len, (jbyte *)native_data->ptr);
+    return jArray;
+}
+
+/**
+ * Converts a Native aws_byte_cursor to a Java ByteBuffer Object
+ */
+jobject aws_jni_byte_buffer_copy_from_cursor(JNIEnv *env, const struct aws_byte_cursor *native_data) {
+    assert(env);
+    jbyteArray jArray = aws_jni_byte_array_from_cursor(env, native_data);
+    jobject jByteBuffer = (*env)->CallStaticObjectMethod(env, s_java_byte_buffer.cls, s_java_byte_buffer.wrap, jArray);
+
+    // Set the Buffer Limit (the max allowed element to read/write)
+    (*env)->CallObjectMethod(env, jByteBuffer, s_java_byte_buffer.set_limit, (jint)native_data->len);
+
+    return jByteBuffer;
+}
+
+/**
+ * Converts a Native aws_byte_cursor to a Java DirectByteBuffer
+ */
+jobject aws_jni_direct_byte_buffer_from_byte_buf(JNIEnv *env, const struct aws_byte_buf *dst) {
+    jobject jByteBuf = (*env)->NewDirectByteBuffer(env, (void *)dst->buffer, (jlong)dst->capacity);
```

Comment:
This can only be used in cases where you guarantee that the ByteBuffer will be copied or dealt with within the scope of the callback you're calling. If any reference is kept to it (for instance, to re-assemble a large file), it will be pointing at dead memory the moment you leave the native callback. 

You probably want to stick to byte[] and do a copy to guarantee safety.

Reply:
Actually, on second thought, just note that in the comment in the header for this function, that the byte buffer is an alias, not a copy

Reply:
Fixed